Feature/async await

TomCarey opened the pull request.

TomCarey committed:
async: Added async-await methods for API Clients

TomCarey committed:
use node 18

TomCarey committed:
oauth: improvements to security model

TomCarey committed:
fix retry for when there are no parameters

TomCarey committed:
errors: forward all of the error information in an auth failure

TomCarey committed:
security: Adding support for authentication through security clients

karlvr commented on templates/frag/operation.hbs:
```diff
@@ -48,55 +48,79 @@ open func {{{name}}}(_ __request: {{{className (concat name '_' 'request')}}}, {
+            let decodedData = try JSONDecoder().decode({{{defaultContent.nativeType.concreteType}}}.self, from: data)
+            return ._{{{code}}}(decodedData)
+        } catch {
+            throw APIError.invalidResponse(error)
```
Shall we include the `response` and the `data` with this error like we do for `unexpectedResponse`? Maybe during development you'd like to examine them when this occurs? I think it means that the server sent something undocumented... which happens.

karlvr commented on templates/frag/operation.hbs:
```diff
@@ -48,55 +48,79 @@ open func {{{name}}}(_ __request: {{{className (concat name '_' 'request')}}}, {
+            if !didAuthenticate {
+                do {
+                    {{#each schemes}}
+                    try await securityClient.authenticationFailed(securityScheme: .{{{scheme.name}}}, scopes: nil)
```
I wonder about renaming this function to describe what it's going to do rather than the situation that occurred to trigger it being called. `handleAuthenticationFailed` would be classic, but perhaps it could be `reauthenticate`... I think that's what we're doing.

karlvr replied:
We should have a value for `scopes` here... I think we can report the scopes in this requirement? We'll need to make that clear in the func signature or documentation that those are the scopes that are required for the endpoint that requires reauthentication... so may not be super useful... but perhaps an implementation would add those scopes to the ones it thinks it already has, for example.

karlvr commented on templates/frag/operation.hbs:
```diff
@@ -48,55 +48,79 @@ open func {{{name}}}(_ __request: {{{className (concat name '_' 'request')}}}, {
             }
+            {{/each}}
+            {{/if}}
+            {{#if securityRequirements.optional}}
```
You could change `#if` to `#unless` here and you don't need the `else` clause.

karlvr commented on templates/frag/operation.hbs:
```diff
@@ -48,55 +48,79 @@ open func {{{name}}}(_ __request: {{{className (concat name '_' 'request')}}}, {
+            {{#if securityRequirements.optional}}
+            {{else}}
+            if (!didAuthenticate) {
+                throw lastError ?? APIError.authenticationPrevented
```
I think `securityClientNotFound` might be better? Or is this actually a `securitySchemeNotConfigured`? I think we only get here if there isn't a `SecurityClient` configured for the `securityScheme`.

TomCarey replied:
not quite, in theory lastError should never be null in this situation. perhaps we ought to bang it or `if let` it

karlvr commented on templates/frag/operation.hbs:
```diff
@@ -150,6 +174,9 @@ open func {{{name}}}Request({{{_params}}}) throws -> URLRequest {
 {{/with}}
 {{/if}}
+
+    {{>frag/requestSecurity}}
```
I wonder if this should also be after the addition of headers... so it's basically the final word on the subject of the request except for...?

karlvr commented on templates/frag/operation.hbs:
```diff
@@ -150,6 +174,9 @@ open func {{{name}}}Request({{{_params}}}) throws -> URLRequest {
+
     localVarHeaderParameter.forEach { item in localVarRequest.addValue(item.value!, forHTTPHeaderField: item.name) }
 
     if let localVarPrepareRequestBlock = self.configuration.prepareRequestBlock {
```
Maybe we should rename `prepareRequestBlock` to `finaliseRequestBlock`... after all _it_ is the final word on the request, whereas `prepare` makes it sound like it's early.

karlvr commented on templates/frag/requestSecurity.hbs:
```diff
@@ -1,45 +1,27 @@
-        localVarHeaderParameter.set("Authorization", "{{capitalize scheme.scheme}} \(localVarAuthorizationValue)")
-    }
+{{/each}}
+{{#if securityRequirements.optional}}
```
`#unless`

karlvr commented on templates/security/SecurityScheme.swift.hbs:
```diff
@@ -1,6 +1,47 @@
 /** The security schemes declared by the API specification. */
+import Foundation
+
 public enum SecurityScheme: String {
```
Let's add a comment

```swift
/// The security schemes documented in the API specification.
```

karlvr commented on templates/support/Configuration.swift.hbs:
```diff
@@ -5,29 +5,16 @@
 
 public typealias ConfigurationAPIKeyBlock = ((_ name: String) -> String?)
-public typealias ConfigurationAuthorizationBlock = ((_ name: String, _ scopes: [String]?) -> String?)
+public typealias ConfigurationAuthorizationBlock = ((_ name: String, _ scopes: [String]?, _ request: URLRequest) -> URLRequest)
```
Instead of `String` can't we use the `SecurityScheme` type?

TomCarey replied:
we have deleted it

karlvr commented on templates/support/Configuration.swift.hbs:
```diff
@@ -5,29 +5,16 @@
 import Foundation
 
 public typealias ConfigurationAPIKeyBlock = ((_ name: String) -> String?)
```
Can we delete this `typealias`?

TomCarey replied:
we can delete most of these actually.

karlvr commented on templates/support/Configuration.swift.hbs:
```diff
@@ -5,29 +5,16 @@
-    public var password: String?
 
     /**
      * Return the authorization value to use for the named security scheme and scopes.
```
```swift
/** The handler for all security requests. */
```

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+//  {{>frag/generatedBy}}
+//
+
+/** A token manager that will group refresh attempts to prevent concurrent token requests */
```
This comment should be on the `actor` I think.

Let's change the comment to:

```swift
/** An OAuth access token manager that manages using and refreshing the access token, including handling concurrent requests and refreshes. */
```

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+
+actor OAuthAccessTokenManager {
+    
+    var accessToken: OAuthToken? {
```
Perhaps this type should be `OAuthAccessToken`... I think "access token" is the right terminology for this thing.

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+    }
+    
+    private let refreshTokenURL: URL?
+    private let automaticallyRefreshTokensThisManySecondsEarly: TimeInterval?
```
Do we need to have "seconds" in this name everywhere when it's a `TimeInterval`? I think a `TimeInterval` is explicit about its units? And we can probably remove the word "automatically" and just call it "preemptiveRefresh" or "preemptiveAccessTokenRefresh"? Or something like that?

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+                do {
+                    try await self.refreshToken()
+                } catch {
+                    print("An error ocurred in the automatic refresh: \(error)")
```
This feels like it's only for debugging... our assumption is that the app will then use an API and get the refresh failure reported in a thrown error that they handle eh? So I guess it's OK to have it here? Maybe we can use the new OSLog thing so it looks a little less like it was just left behind?

TomCarey replied:
The logger needed would look something like this 
```swift
@available(macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0, *)
private let logger = Logger(subsystem: "{{@root.generatorClass}}", category: "oauth")
```
do you have a better subsystem name?

karlvr replied:
Perhaps `@root.info.title`

TomCarey replied:
`@root.info.title` is empty, perhaps it's not being ingested properly from the yaml? Furthermore, apple suggests a reverse domain style. Perhaps we could have a name like `swift-client-generator.{{@root.info.title}}` might even want to convert that title to a classname version of it?

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+/** A token manager that will group refresh attempts to prevent concurrent token requests */
+import Foundation
+
+actor OAuthAccessTokenManager {
```
I wonder how much of this should be `public` so we can make use of it outside of the API module... I feel like quite a bit of it should be? We might want to interact with the access token etc from app code.

TomCarey replied:
We don't need the actor to be public for that. I have exposed the access token on the security client via an async function

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+        })
+    }
+    
+    func setAccessToken(_ token: OAuthToken) async throws {
```
This is a bit weird to have a `setAccessToken` and a visible `accessToken` property isn't it? This one is async... do we override it sometimes to do some async stuff? I think then we should have a `private` setter on the `accessToken` ivar.

TomCarey replied:
it's weird but from my understanding a consequence of using an actor. Actors prevent using the normal setter

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+    
+}
+
+public struct OAuthToken : Decodable, Equatable {
```
This is used from a few spots so let's move this into its own file eh?

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+            throw APIError.authenticationPrevented
+        }
+        
+        if let t = refreshTask {
```
Let's document this:

```swift
// Wait for an existing refresh task, if there is one
```

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+            default:
+                throw APIError.authenticationFailed(result.response, data: result.data)
+            }
+            
```
We could tidy up this blank line.

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+            try await refreshToken()
+        }
+        
+        if let t = refreshTask {
```
Let's document:

```swift
// Wait for an existing refresh task, if there is one
```

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+    func authorize(securityScheme: SecurityScheme, scopes: [String]?, request: URLRequest) async throws -> URLRequest
+}
+
+/** Delegates security requiements to the appropriate SecurityClient. */
```
Typo on "requirements"

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+}
+
+/** Delegates security requiements to the appropriate SecurityClient. */
+public class SecurityClientController: SecurityClient {
```
I'm feeling like it's cool but a little confusing that we've got `SecurityClient` as the protocol for this multiplexer and for our individual implementations... it makes sense technically but I think it's a little confusing: "Do I put my security client into the `Configuration` directly or..."

Perhaps we should _default_ the `securityClient` on `Configuration` to a `SecurityClientController` so it's a little more obvious that you add your own `SecurityClient` to the controller, not directly to the `Configuration`.

TomCarey replied:
I'm not sure about this. It seems like we could modify the constructor for Configuration to accept a type of `SecurityClientController` but when you read the value from the configuration it will be a `SecurityClient` perhaps it just needs two types even though they effectively look identical

karlvr replied:
We agreed to decide to leave it how it is for the moment and see how the code feels. I want to help someone generating code to understand how it works.

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+        if let client = clients[securityScheme] {
+            return try await client.authorize(securityScheme: securityScheme, scopes: scopes, request: request)
+        } else {
+            return request
```
Why doesn't this throw `APIError.securitySchemeNotConfigured(securityScheme)`? The API spec says the request needs to be authorised so I think we should throw an error if we're not able to do so.

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+        self.clients[securityScheme] = client
+    }
+
+    public init(clients: [SecurityScheme : SecurityClient] = [:]) {
```
Is this normal to put the constructor at the bottom of the class? I thought they normally went near the top!

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+    return refreshRequest
+}
+
+public class AbstractOAuthFlowClient: SecurityClient {
```
I think it will be nicer to break these into their own named files.

karlvr commented on templates/support/URLSession+Api.swift.hbs:
```diff
@@ -8,12 +8,27 @@ extension Foundation.URLSession {
+        do {
+            let (data, response) = try await apiSession.data(for: request)
+            guard let httpResponse = response as? HTTPURLResponse else {
+                fatalError("Expected non-nil HTTPURLResponse from URLRequest")
```
This is a little funny in a library like this? We've got an opportunity to report an error so perhaps we can report a `systemError` with a message or something like that?

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+            if let cookie = HTTPCookie(properties: cookieProperties) {
+                let headers = HTTPCookie.requestHeaderFields(with: [cookie])
+                headers.forEach { key, value in
+                    request.setValue(value, forHTTPHeaderField: key)
```
I think we need to use `addValue` rather than `setValue` here. There might be multiple cookies and they all use the header name `"Cookie"`.

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+    
+    func setRefreshToken(_ token: String) async throws {
+        try await tokenManager.setRefreshToken(token)
+        try await tokenManager.refreshToken()
```
I feel like this is a bit enigmatic... it makes sense inside the following `authenticate` method... can this code be in the `authenticate` method instead?

Perhaps we shouldn't have this method at all and we should instead have an `OAuthExistingSessionClient` that takes a refresh token and continues the session... it's a bit strange that all of our flows support this, when our use-case for this is just resuming a session. Do we need to care what flow was used to create the session? I don't think we do... so then the app doesn't need to try to reuse the right flow client when resuming a session (such as by getting the refresh token out of the keychain).

TomCarey replied:
whether or not a refresh token works depends on the refreshUrl supplied by the pre defined flow. It seems to me that the consumer of the API will need to know which flow they were on before supplying a refresh token.

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+import Foundation
+
+public protocol SecurityClient {
+    /** An API request received a 401 response */
```
Let's expand this documentation, taking into account my earlier renaming suggestion if appropriate. I think we want to document that the session will have been reauthenticated when this method returns, otherwise it throws an error.

karlvr commented on templates/support/SecurityClient.swift.hbs:
```diff
@@ -0,0 +1,361 @@
+    /** An API request received a 401 response */
+    func authenticationFailed(securityScheme: SecurityScheme, scopes: [String]?) async throws
+    
+    /** An authorization header to the request */
```
```swift
/** Modify the request to authorise it for the required SecurityScheme and scopes. */
```

Perhaps we should move the `request` parameter to the first position? That might _read_ better.

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,126 @@
+    
+    func refreshToken() async throws {
+        guard let refreshTokenURL = refreshTokenURL else {
+            throw APIError.authenticationPrevented
```
It's almost like here we're simply saying "you're no longer authenticated" ? Maybe we need to look at each place where this func is called and see what a failure here means to that calling method.

TomCarey replied:
I'm not sure the distinction is necessary? if we don't have a url then we can't refresh, that doesn't mean we are no longer authenticated unless you also have an expired token (which obviously would be the common path) but we can refresh pre-emptively (or try to) and we'd get this error.

TomCarey committed:
errors: add response and data to errors for improved debugging

TomCarey committed:
securityclient: improved naming for clearer intention

TomCarey committed:
logic: clearer code using unless

TomCarey committed:
scopes: forward the required scopes to the authentication endpoint

TomCarey committed:
error: the lastError error should be nonnull by the time we get to this line

TomCarey committed:
documentation: improved documentation

TomCarey committed:
cleanup: deleted unused code

TomCarey committed:
cleanup: renamed prepare block for clarity to finalizeBlock

TomCarey committed:
cleanup: renamed OAuthToken to OAuthAccessToken for consistency

TomCarey committed:
cleanup: Rename refresh token time for clarity making use of the type to descibe the unit

TomCarey committed:
logging: use os log when available

TomCarey committed:
module: expose mechanism to fetch the current access token

TomCarey committed:
refactor: Move OAuthAccessToken into its own file

TomCarey committed:
documentation: explaining the refresh task

TomCarey committed:
cleanup: remove blank line

TomCarey committed:
documenation: added context and fixed typos

TomCarey committed:
errors: throw on missing scheme rather than silently fail

TomCarey committed:
refactor: code tidy up

TomCarey committed:
refactor: Extract classes into separate files

TomCarey committed:
errors: throw an error when handling an unexpected response

TomCarey committed:
apikey: fix cookie flow to add values rather than clobber them

TomCarey committed:
documentation: added documentation

TomCarey committed:
cleanup: reorder parameters for clarity

TomCarey committed:
tidy: removed excess methods for setting tokens

TomCarey committed:
logs: use the API title for the logger subsystem

TomCarey committed:
securityclient: refactor helper constructors to remove unusual error pattern in authorization

TomCarey committed:
errors: change authenticationPrevented to notAuthenticated

karlvr committed:
Merge branch 'main' into feature/async-await

# Conflicts:
#	pnpm-lock.yaml

karlvr committed:
test: add OAuth2 test spec

karlvr committed:
Merge branch 'main' into feature/async-await

karlvr committed:
Target macOS 10.15 as we now require async await

karlvr committed:
Fix handling of security when no security schemes are documented

karlvr committed:
chore: remove whitespace

karlvr committed:
Fix handling of request security when there are no schemes documented

karlvr commented on templates/support/APIError.swift.hbs:
```diff
@@ -0,0 +1,20 @@
+
+import Foundation
+
+public enum APIError: Error {
```
Let's add documentation for `APIError` itself...

```swift
/// Errors that occur while using the API
```

karlvr commented on templates/support/APIError.swift.hbs:
```diff
@@ -0,0 +1,20 @@
+import Foundation
+
+public enum APIError: Error {
+    /// The authentication could not be attempted
```
I think this error is used if we try to refresh a token and we have no `refreshTokenURL`, so we literally cannot attempt a refresh, and that's a compile-time problem (we didn't have that URL). But also if we try to refresh and we don't have a refresh token... which is a runtime problem. Or if we simply don't have an access token... a common runtime problem if we try to call an API that requires auth but we don't have it.

The name `notAuthenticated` is good, but I think the docs should say:

```swift
/// The session is not authenticated
```

I'm not sure about the use of the word "session"... but what is it? User? Not necessarily, we might be using client credentials. "We"? Who even are we? The Security client is not authenticated?

karlvr commented on templates/support/APIError.swift.hbs:
```diff
@@ -0,0 +1,20 @@
+    /// The authentication request failed
+    case authenticationFailed(_ response: HTTPURLResponse, data: Data)
+
+    case invalidResponse(_ error: Error, response: HTTPURLResponse, data: Data)
```
```swift
/// A response was received that could not be decoded
```

karlvr commented on templates/support/APIError.swift.hbs:
```diff
@@ -0,0 +1,20 @@
+
+    case invalidResponse(_ error: Error, response: HTTPURLResponse, data: Data)
+
+    case systemError(_ error: Error)
```
```swift
/// A system error occurred, such as a networking error
```

karlvr commented on templates/support/APIError.swift.hbs:
```diff
@@ -0,0 +1,20 @@
+
+    case systemError(_ error: Error)
+
+    case securitySchemeNotConfigured(_ securityScheme: SecurityScheme)
```
```swift
/// The API requested authentication using a `SecurityScheme` that has not been configured in the `SecurityClientController`
```

This error is a bit weird. It's specific to `SecurityClientController`... It's handier than just reporting `notAuthenticated` so I guess we keep it... but it's weird. I think I'm in the java mindset where you probably want to catch and handle the various exceptions that are thrown, whereas we'll probably never want to catch this... it just reports what happened nicely.

karlvr commented on templates/support/APIError.swift.hbs:
```diff
@@ -0,0 +1,20 @@
+    case systemError(_ error: Error)
+
+    case securitySchemeNotConfigured(_ securityScheme: SecurityScheme)
+    case unexpectedResponse(_ response: URLResponse, data: Data)
```
```swift
/// An response was received with an undocumented status code
```

We need to change our use of this error in `URLSession+Api.swift` when we don't get an `HTTPURLResponse`... that's such a weird event maybe we could just not mention it.

karlvr commented on templates/support/Configuration.swift.hbs:
```diff
@@ -4,30 +4,12 @@
-public typealias ConfigurationAPIKeyBlock = ((_ name: String) -> String?)
-public typealias ConfigurationAuthorizationBlock = ((_ name: String, _ scopes: [String]?) -> String?)
-public typealias ConfigurationPrepareRequestBlock = ((_ request: inout URLRequest) -> Void)
+public typealias ConfigurationFinalizeReleaseBlock = ((_ request: inout URLRequest) -> Void)
```
`ConfigurationFinalizeRequestBlock`?

karlvr commented on templates/support/BasicAuthenticationSecurityClient.swift.hbs:
```diff
@@ -0,0 +1,35 @@
+        
+    }
+    
+    public func authorize(request: URLRequest, securityScheme: SecurityScheme, scopes: [String]?) async throws -> URLRequest {
```
Do you think we should have made this `inout URLRequest`? It seems pretty obvious that's what they all need to do.

karlvr commented on templates/support/Configuration.swift.hbs:
```diff
@@ -40,27 +22,21 @@ public struct Configuration {
     public var responseQueue: DispatchQueue?
 
-    public var prepareRequestBlock: ConfigurationPrepareRequestBlock?
+    public var finalizeReleaseBlock: ConfigurationFinalizeReleaseBlock?
```
`finalizeRequestBlock`?

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,131 @@
+                    if #available(macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
+                        logger.error("An error ocurred in the automatic refresh: \(error)")
+                    } else {
+                        print("An error ocurred in the automatic refresh: \(error)")
```
I think we should probably identify ourselves here... how about `{{@root.info.title}}: OAuth: An error...`, and fix speeling of occurred.

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,131 @@
+                    try await self.refreshToken()
+                } catch {
+                    if #available(macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
+                        logger.error("An error ocurred in the automatic refresh: \(error)")
```
```swift
"An error occurred in the automatic OAuth access token refresh: \(error)"
```

Note two 'c's in 'occurred' too.

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,131 @@
+            throw APIError.notAuthenticated
+        }
+        
+        /// Wait for an existing refresh task, if there is one
```
I think it's just two `/`s in this type of comment. The three `/` comments are for method and type documentation,

karlvr commented on templates/support/OAuthAccessTokenManager.swift.hbs:
```diff
@@ -0,0 +1,131 @@
+        }
+        
+        /// Wait for an existing refresh task, if there is one
+        if let t = refreshTask {
```
If we have an access token that we believe is still valid, then we shouldn't wait for a refresh task eh? That task might be our preemptive refresh. So we don't want this block at all here... if the token isn't expired, then we try to use it.

TomCarey replied:
That might introduce a race condition where the refresh might invalidate the token so the request might get a 401 response and then force another refresh.

karlvr committed:
docs: tweak documentation

karlvr commented on templates/support/OAuthAuthorizationCodeFlowClient.swift.hbs:
```diff
@@ -0,0 +1,25 @@
+        
+        super.init(clientId: clientId, clientSecret: clientSecret, refreshURL: refreshURL, preemptiveAccessTokenRefresh: autoRefreshInterval)
+    }
+    
```
This client isn't done yet?

karlvr commented on templates/support/AbstractOAuthFlowClient.swift.hbs:
```diff
@@ -0,0 +1,72 @@
+        self.tokenManager = OAuthAccessTokenManager(refreshTokenURL: refreshURL, preemptiveAccessTokenRefresh: autoRefreshInterval)
+    }
+    
+    public func authenticate(refreshToken: String) async throws {
```
```swift
/// Authenticate the security client using a refresh token.
```

karlvr commented on templates/support/OAuthCredentialsFlowClient.swift.hbs:
```diff
@@ -0,0 +1,54 @@
+        super.init(clientId: clientId, clientSecret: clientSecret, refreshURL: refreshURL, preemptiveAccessTokenRefresh: autoRefreshInterval)
+    }
+    
+    public func authenticate(scopes: [String]?, additionalParams params: [String: String]? = nil) async throws {
```
```swift
/// Authenticate the security client, requesting the given scopes.
```

karlvr commented on templates/support/OAuthPasswordFlowClient.swift.hbs:
```diff
@@ -0,0 +1,48 @@
+        super.init(clientId: clientId, clientSecret: clientSecret, refreshURL: refreshURL, preemptiveAccessTokenRefresh: autoRefreshInterval)
+    }
+    
+    public func authenticate(username: String, password: String, scopes: [String]?, additionalParams params: [String: String]? = nil) async throws {
```
```swift
/// Authenticate the security client using username and password credentials, requesting the given scopes.
```

TomCarey committed:
oauth: Allow access token changes to be observed

karlvr commented on templates/support/SecurityClientController.swift.hbs:
```diff
@@ -0,0 +1,41 @@
+
+ - Note: Specific `SecurityClient` objects can be inserted or replaced dynamically to accommodate user behavior changes, such as signing in or signing out.
+*/
+public class SecurityClientController: SecurityClient {
```
This name is feeling quite generic now given that most people simply might not use it... as most APIs have a single scheme. Perhaps the name should be `MultiSecurityClientSupport` even.

TomCarey replied:
Discussed internally, we'll keep it as is

karlvr committed:
oauth: document what’s remaining to do on the authorization code flow

karlvr committed:
Remove doomed attempt to support security when none is specified

karlvr committed:
chore: fix warnings

TomCarey committed:
typo: Change Release to Request

TomCarey committed:
documentation: improved APIError

TomCarey committed:
chore: improved logging

karlvr committed:
chore: tidy indentation

TomCarey committed:
docs: tidy up

TomCarey committed:
docs: added docs to authenticate

karlvr committed:
chore: tidy comments

karlvr committed:
Move security-related files into security folder

karlvr committed:
Tidy generated OAuth flow clients

So we can tolerate multiple OAuth schemes in one spec

karlvr committed:
Use inout on authorize(request…) func

karlvr committed:
Pass failed request to reauthenticate

So SecurityClient implementation can manage races between authentication failures.

karlvr committed:
chore: rename misnamed file

karlvr committed:
chore: warnings

karlvr committed:
Only refresh an access token if it is still current

Don’t block authorize operations while we’re refreshing, unless the access token has already expired.

karlvr committed:
Fix references to security scheme name

They need to use the identifier helper to match what’s in SecurityScheme.swift’s enum

TomCarey committed:
errors: clean up error handling on operations

TomCarey committed:
oauth: added revocation capability

karlvr commented on templates/frag/operation.hbs:
```diff
@@ -40,63 +36,86 @@ public enum {{className name}}Result {
 @available(*, deprecated)
 {{/if}}
-open func {{{name}}}(_ __request: {{{className (concat name '_' 'request')}}}, {{#if requestBody.nativeType}}{{{requestBody.name}}}: {{{requestBody.nativeType}}}{{#unless requestBody.required}} = nil{{/unless}}, {{/if}}responseQueue: DispatchQueue? = nil, completion: @escaping ((_ result: {{className name}}Result) -> Void)) {
+open func {{{name}}}(_ __request: {{{className (concat name '_' 'request')}}}, {{#if requestBody.nativeType}}{{{requestBody.name}}}: {{{requestBody.nativeType}}}{{#unless requestBody.required}} = nil{{/unless}}, {{/if}}responseQueue: DispatchQueue? = nil, completion: @escaping ((_ result: Result<{{className name}}Result, Error>) -> Void)) {
```
If we're still using `APIError.systemError` then perhaps the error type can by `APIError` or is that meaningless and `Error` is fine... errors aren't type safe ?

TomCarey replied:
removed the wrapped system error

karlvr commented on templates/frag/operation.hbs:
```diff
@@ -27,10 +27,6 @@ public enum {{className name}}Result {
     {{/if}}
     case _{{{code}}}{{#if defaultContent.nativeType}}(_ value: {{{defaultContent.nativeType}}}){{/if}}
```
Ooh! We need to omit `401` if it's in here, as we'll never respond with that status code.

TomCarey committed:
errors: exclude 401 errors from the valid responses

It would never be returned anyway as there is custom handling for 401

TomCarey committed:
errors: remove the wrapping of system errors

It was causing wasteful unwrapping

TomCarey committed:
oauth: fix bug with token expiry check

TomCarey committed:
requests: added equatable and hashable to requests

So that consumers can compare requests

TomCarey committed:
chore: use fully qualified types

TomCarey committed:
async: added missing async operation using request structs

TomCarey committed:
docs: remove markdown call which was adding p tags that we dont want
